### ISSUE TITLE
Conditional buffer types

### ISSUE THREAD

**Issue by binarylogic** [vectordotdev/vector#10384]
This issue should be considered in the broader context of #10281 since it may not be applicable based on our solution for that.

---

As an alternative to #10327 and #10326, we should consider conditional buffer types. This solves the sink-level fan-in use case where data is coming from different sources with different levels of upstream durability.

For example, take a graph where a single `datadog_logs` source is receiving events from a `datadog_agent` source and a `kafka` source:

```
datadog_agent --\
                 ------ datadog_logs
        kafka --/
```

The Datadog Agent is designed to shed load in order to protect the reliability of the primary service that it cohabitates the node with. Therefore, back pressure to the `datadog_agent` source is inappropriate since data will be dropped; disk buffers should be used to offer durability. Alternatively, back pressure is appropriate for the `kafka` source since Kafka is designed for buffering data in a durable fashion. We would be downgrading durability by storing Kafka data in a disk buffer.

To solve this, you could imagine conditional buffer types at the sink-level, where `datadog_agent` data buffers in a disk buffer and `kafka` data buffers in an in-memory buffer. This switch could be driven by event metadata. End-to-end acknowledgements would be enabled for both sources ensuring data reaches the next stage of durability reliably.

### 1. How does this deal with ordering?

Because sources and buffer types are consistent, we will not be violating the original ordering. Data going through the disk buffer may be slower, but it won't be out of order.

**Comment by jszwedko** [comment]
Do you have thoughts about this vs. source-based buffers? Naively source-based buffers seem easier to reason about in this multi-source one sink setup.

**Comment by binarylogic** [comment]
I tend to agree. I'm just throwing it all in a pile so we can reason about them when we get further along into considering buffer defaults.
